### PR TITLE
Nonzero performance improvement

### PR DESCRIPTION
**Pull request by ifedan**
Based on tests, if we will assume that the result will have same size as original tensor and resize it after calculation of indices, the performance will be better for large sizes (>= 100)

image

### CODE REVIEW

**Comment by ifedan** [comment]
```python
import timeit
import itertools
import statistics

for i in [10, 100, 10000, 100000, 1000000]:
    setup_code = '''import torch; w = (torch.rand({0}) < 0.5).double()'''.format(i)
    test_code = '''torch.nonzero(w);'''
    result = timeit.repeat(setup = setup_code,stmt = test_code,repeat = 3,number = 100)
    mean = statistics.mean(result)
    std = statistics.stdev(result)
    print('''size = {0} time = {1}'''.format(i, mean))
```

**Comment by VitalyFedyunin** [comment]
Conceptually interesting, but you are doing 2 costy `resize_` operations instead of one.  It seems to work fine on bigger inputs, but I'm afraid it is not going to work well when number of non-zero elements is significantly lower than number of elements.

File: aten/src/ATen/native/cpu/NonzeroKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  /* Second pass populates subscripts */
+  auto subscript_data = subscript.data_ptr<int64_t>();
+  auto subscript_strides = subscript.strides();
+  auto stride0 = subscript_strides[0] - subscript_strides[1] * self.dim();
+  auto stride1 = subscript_strides[1];
+  int64_t numel = 0;
+  auto iter = TensorIterator();
+  iter.add_input(self);
+  iter.build();
+  cpu_serial_kernel(iter, [&](scalar_t a) {
```
Comment (resistor) [inline aten/src/ATen/native/cpu/NonzeroKernel.cpp]: This should probably be capturing by value rather than by reference?

**Comment by ifedan** [comment]
integrated with https://github.com/pytorch/pytorch/pull/27217

**Pull request closed without merging**